File: etl/sdtm_automation/generate_define_xml.R
#!/usr/bin/env Rscript
#=============================================================================
# DEFINE-XML 2.1 GENERATOR FOR SDTM DOMAINS
# Creates FDA-required Define-XML metadata document
# Required for all FDA submissions since March 15, 2023
#=============================================================================

library(xportr)
library(dplyr)
library(readr)
library(here)
library(logger)
library(purrr)
library(xml2)

log_info("Generating Define-XML 2.1 for SDTM domains")

# Study-level metadata
study_metadata <- list(
  studyid = "RECIST-DEMO",
  study_name = "RECIST Response Assessment Demonstration Study",
  protocol_name = "RECIST-DEMO-001",
  sponsor = "Demo Pharmaceutical Inc.",
  define_version = "2.1",
  standard = "SDTM",
  standard_version = "3.4",
  ig_version = "3.4",
  creation_date = Sys.Date()
)

# Domain metadata - describes each SDTM domain
domain_metadata <- tribble(
  ~domain, ~label, ~description, ~class, ~purpose,
  "DM", "Demographics", "Subject-level demographic and study participation information", "SPECIAL PURPOSE", "Tabulation",
  "TU", "Tumor Identification", "Baseline tumor inventory for target and non-target lesions", "FINDINGS", "Tabulation",
  "TR", "Tumor Results", "Longitudinal tumor measurements per RECIST criteria", "FINDINGS", "Tabulation",
  "RS", "Disease Response", "Overall response assessment per RECIST 1.1", "FINDINGS", "Tabulation",
  "AE", "Adverse Events", "Adverse events and safety observations", "EVENTS", "Tabulation",
  "EX", "Exposure", "Study treatment exposure and dosing records", "INTERVENTIONS", "Tabulation",
  "SV", "Subject Visits", "Subject visit attendance and completion status", "EVENTS", "Tabulation",
  "DS", "Disposition", "Subject disposition and study completion status", "EVENTS", "Tabulation",
  "CM", "Concomitant Medications", "Non-study medications taken during trial", "INTERVENTIONS", "Tabulation",
  "MH", "Medical History", "Pre-existing medical conditions before study enrollment", "EVENTS", "Tabulation",
  "LB", "Laboratory Tests", "Laboratory test results including hematology and chemistry", "FINDINGS", "Tabulation",
  "VS", "Vital Signs", "Vital signs measurements including blood pressure and pulse", "FINDINGS", "Tabulation",
  "EG", "ECG Tests", "Electrocardiogram measurements for cardiac safety", "FINDINGS", "Tabulation",
  "PE", "Physical Examination", "Physical examination findings by body system", "FINDINGS", "Tabulation",
  "QS", "Questionnaires", "Patient-reported outcomes and quality of life assessments", "FINDINGS", "Tabulation"
)

# Variable-level metadata template for common SDTM variables
common_variables <- tribble(
  ~variable, ~label, ~type, ~length, ~role, ~cdisc_notes,
  "STUDYID", "Study Identifier", "text", 200, "Identifier", "Unique identifier for the study",
  "DOMAIN", "Domain Abbreviation", "text", 2, "Identifier", "Two-character domain code",
  "USUBJID", "Unique Subject Identifier", "text", 200, "Identifier", "Unique subject identifier across all studies",
  "SUBJID", "Subject Identifier for the Study", "text", 200, "Topic", "Subject identifier within the study",
  "VISITNUM", "Visit Number", "integer", 8, "Timing", "Numeric visit identifier",
  "VISIT", "Visit Name", "text", 200, "Synonym Qualifier", "Protocol-defined visit name",
  "VISITDY", "Planned Study Day of Visit", "integer", 8, "Timing", "Planned study day of visit relative to RFSTDTC"
)

# Function to create variable metadata for a specific domain
create_domain_variables <- function(domain_code) {
  # This is a simplified version - in production, read from actual XPT files
  # or metadata specifications
  
  base_vars <- common_variables %>%
    mutate(domain = domain_code)
  
  # Add domain-specific variables based on domain type
  domain_specific <- switch(domain_code,
    "DM" = tribble(
      ~variable, ~label, ~type, ~length, ~role, ~cdisc_notes,
      "ARM", "Description of Planned Arm", "text", 200, "Synonym Qualifier", "Planned treatment arm",
      "ARMCD", "Planned Arm Code", "text", 20, "Topic", "Short code for planned arm",
      "ACTARM", "Description of Actual Arm", "text", 200, "Synonym Qualifier", "Actual treatment arm",
      "ACTARMCD", "Actual Arm Code", "text", 20, "Topic", "Short code for actual arm",
      "AGE", "Age", "integer", 8, "Record Qualifier", "Age at study entry",
      "AGEU", "Age Units", "text", 10, "Variable Qualifier", "Units for age (YEARS)",
      "SEX", "Sex", "text", 1, "Record Qualifier", "Sex of subject (M/F)",
      "RACE", "Race", "text", 200, "Record Qualifier", "Race of subject",
      "ETHNIC", "Ethnicity", "text", 200, "Record Qualifier", "Ethnicity",
      "RFSTDTC", "Subject Reference Start Date/Time", "datetime", 20, "Record Qualifier", "Reference start date for subject",
      "RFENDTC", "Subject Reference End Date/Time", "datetime", 20, "Record Qualifier", "Reference end date for subject",
      "DTHDTC", "Date/Time of Death", "datetime", 20, "Record Qualifier", "Date/time of death if applicable",
      "DTHFL", "Subject Death Flag", "text", 1, "Record Qualifier", "Y if subject died"
    ),
    "AE" = tribble(
      ~variable, ~label, ~type, ~length, ~role, ~cdisc_notes,
      "AESEQ", "Sequence Number", "integer", 8, "Identifier", "Sequence number for AE",
      "AETERM", "Reported Term for the Adverse Event", "text", 200, "Topic", "Verbatim AE term",
      "AEDECOD", "Dictionary-Derived Term", "text", 200, "Synonym Qualifier", "MedDRA preferred term",
      "AESOC", "Primary System Organ Class", "text", 200, "Grouping Qualifier", "MedDRA SOC",
      "AESEV", "Severity/Intensity", "text", 20, "Record Qualifier", "Severity (MILD/MODERATE/SEVERE)",
      "AESER", "Serious Event", "text", 1, "Record Qualifier", "Y if serious AE",
      "AEREL", "Causality", "text", 50, "Record Qualifier", "Relationship to study drug",
      "AEACN", "Action Taken with Study Treatment", "text", 50, "Record Qualifier", "Action taken",
      "AEOUT", "Outcome of Adverse Event", "text", 50, "Record Qualifier", "AE outcome",
      "AESTDTC", "Start Date/Time of Adverse Event", "datetime", 20, "Timing", "AE start date",
      "AEENDTC", "End Date/Time of Adverse Event", "datetime", 20, "Timing", "AE end date"
    ),
    "LB" = tribble(
      ~variable, ~label, ~type, ~length, ~role, ~cdisc_notes,
      "LBSEQ", "Sequence Number", "integer", 8, "Identifier", "Sequence number",
      "LBTESTCD", "Lab Test or Examination Short Name", "text", 8, "Topic", "Lab test code",
      "LBTEST", "Lab Test or Examination Name", "text", 200, "Synonym Qualifier", "Lab test name",
      "LBCAT", "Category for Lab Test", "text", 200, "Grouping Qualifier", "HEMATOLOGY/CHEMISTRY",
      "LBSPEC", "Specimen Type", "text", 200, "Record Qualifier", "Specimen type",
      "LBORRES", "Result or Finding in Original Units", "text", 200, "Result Qualifier", "Original result",
      "LBORRESU", "Original Units", "text", 40, "Variable Qualifier", "Original units",
      "LBSTRESC", "Character Result/Finding in Std Format", "text", 200, "Result Qualifier", "Standardized result",
      "LBSTRESN", "Numeric Result/Finding in Standard Units", "float", 8, "Result Qualifier", "Numeric result",
      "LBSTRESU", "Standard Units", "text", 40, "Variable Qualifier", "Standard units",
      "LBSTNRLO", "Reference Range Lower Limit in Std Units", "float", 8, "Variable Qualifier", "Normal range low",
      "LBSTNRHI", "Reference Range Upper Limit in Std Units", "float", 8, "Variable Qualifier", "Normal range high",
      "LBNRIND", "Reference Range Indicator", "text", 20, "Variable Qualifier", "NORMAL/HIGH/LOW",
      "LBBLFL", "Baseline Flag", "text", 1, "Record Qualifier", "Y if baseline",
      "LBDTC", "Date/Time of Specimen Collection", "datetime", 20, "Timing", "Collection date"
    ),
    # Default for other domains
    tribble(
      ~variable, ~label, ~type, ~length, ~role, ~cdisc_notes,
      paste0(domain_code, "SEQ"), "Sequence Number", "integer", 8, "Identifier", "Sequence number"
    )
  )
  
  bind_rows(base_vars, domain_specific %>% mutate(domain = domain_code))
}

# Generate Define-XML 2.1 structure
generate_define_xml <- function() {
  log_info("Creating Define-XML 2.1 structure")
  
  # Create XML root
  define <- xml_new_root(
    "ODM",
    xmlns = "http://www.cdisc.org/ns/odm/v1.3",
    "xmlns:def" = "http://www.cdisc.org/ns/def/v2.1",
    "xmlns:xlink" = "http://www.w3.org/1999/xlink",
    ODMVersion = "1.3.2",
    FileOID = paste0("Define.", study_metadata$studyid, ".", format(Sys.Date(), "%Y%m%d")),
    FileType = "Snapshot",
    CreationDateTime = format(Sys.time(), "%Y-%m-%dT%H:%M:%S"),
    Originator = study_metadata$sponsor
  )
  
  # Add Study section
  study_node <- xml_add_child(define, "Study", OID = study_metadata$studyid)
  
  # Global Variables
  global_vars <- xml_add_child(study_node, "GlobalVariables")
  xml_add_child(global_vars, "StudyName", study_metadata$study_name)
  xml_add_child(global_vars, "StudyDescription", "RECIST Response Assessment Oncology Trial")
  xml_add_child(global_vars, "ProtocolName", study_metadata$protocol_name)
  
  # MetaDataVersion
  metadata_version <- xml_add_child(
    study_node, 
    "MetaDataVersion",
    OID = paste0(study_metadata$studyid, ".SDTM"),
    Name = "SDTM Metadata",
    Description = "SDTM 3.4 Implementation",
    "def:DefineVersion" = "2.1",
    "def:StandardName" = "SDTM-IG",
    "def:StandardVersion" = study_metadata$ig_version
  )
  
  # Add ItemGroupDef for each domain
  log_info("Adding domain definitions")
  
  for (i in seq_len(nrow(domain_metadata))) {
    domain <- domain_metadata[i, ]
    
    item_group <- xml_add_child(
      metadata_version,
      "ItemGroupDef",
      OID = paste0("IG.", domain$domain),
      Name = domain$domain,
      Repeating = "Yes",
      "def:Class" = domain$class,
      "def:Purpose" = domain$purpose,
      "def:Structure" = "One record per subject per event"
    )
    
    xml_add_child(item_group, "Description")
    xml_add_child(xml_children(item_group)[[1]], "TranslatedText", 
                  "xml:lang" = "en", domain$description)
    
    # Add variables for this domain
    domain_vars <- create_domain_variables(domain$domain)
    
    for (j in seq_len(nrow(domain_vars))) {
      var <- domain_vars[j, ]
      xml_add_child(
        item_group,
        "ItemRef",
        ItemOID = paste0("IT.", domain$domain, ".", var$variable),
        OrderNumber = as.character(j),
        Mandatory = if(var$role == "Identifier") "Yes" else "No"
      )
    }
  }
  
  log_info("Define-XML structure created successfully")
  return(define)
}

# Create output directory
dir.create(here("outputs", "define"), recursive = TRUE, showWarnings = FALSE)

# Generate Define-XML
define_xml <- generate_define_xml()

# Write to file
output_path <- here("outputs", "define", "define.xml")
write_xml(define_xml, output_path, options = c("format", "no_declaration"))

log_info("✓ Define-XML 2.1 written: {output_path}")

# Also create a summary report
summary_report <- domain_metadata %>%
  mutate(
    variables_count = map_int(domain, ~nrow(create_domain_variables(.x))),
    status = "Complete"
  )

summary_path <- here("outputs", "define", "define_summary.csv")
write_csv(summary_report, summary_path)

log_info("✓ Define summary written: {summary_path}")

message("\n========================================")
message("Define-XML 2.1 Generation Complete")
message("========================================")
message(sprintf("Domains documented: %d", nrow(domain_metadata)))
message(sprintf("Standard: SDTM-IG v%s", study_metadata$ig_version))
message(sprintf("Define-XML version: %s", study_metadata$define_version))
message(sprintf("Output: %s", output_path))
message("========================================\n")

message("NOTE: This is a template Define-XML generator.")
message("For production use, enhance with:")
message("  - Full variable-level metadata from actual XPT files")
message("  - Controlled terminology references")
message("  - Value-level metadata and codelists")
message("  - Where clauses and comments")
message("  - Origin metadata (CRF pages, derivations)")

File: automation/generate_define_xml.py
#!/usr/bin/env python3
"""
Define-XML Generator using odmlib
Generates CDISC Define-XML v2.1 from study metadata for RECIST 1.1 implementation

Part of CDISC 360i automation pipeline
References:
- odmlib: https://github.com/swhume/odmlib
- Define-XML v2.1: https://www.cdisc.org/standards/data-exchange/define-xml
"""

from odmlib.define_2_1 import model as ODM
import json
from datetime import datetime
from pathlib import Path
import sys

def create_study_metadata():
    """Create ODM Study root with metadata"""
    
    # Create ODM root element
    odm = ODM.ODM(
        FileOID=f"DEFINE-RECIST-{datetime.now().strftime('%Y%m%d')}",
        FileType="Snapshot",
        CreationDateTime=datetime.now().isoformat(),
        ODMVersion="1.3.2",
        Originator="SAS-R Hybrid Pipeline",
        SourceSystem="odmlib + pharmaverse",
        SourceSystemVersion="1.0.0"
    )
    
    # Create study
    study = ODM.Study(
        OID="RECIST-DEMO-001",
        StudyName="RECIST 1.1 Demonstration Study",
        StudyDescription="Oncology trial demonstrating automated SDTM/ADaM generation with pharmaverse",
        ProtocolName="RECIST-DEMO-001"
    )
    
    # Create global variables
    global_vars = ODM.GlobalVariables(
        StudyName="RECIST 1.1 Demo",
        StudyDescription="Automated clinical pipeline demonstration",
        ProtocolName="RECIST-DEMO-001"
    )
    study.GlobalVariables = global_vars
    
    # Create metadata version
    mdv = ODM.MetaDataVersion(
        OID="MDV.RECIST-DEMO.SDTM.1.0",
        Name="SDTM Metadata Version 1.0",
        Description="RECIST 1.1 SDTM implementation with sdtm.oak automation",
        DefineVersion="2.1"
    )
    
    # Add standard information
    mdv.Standard = []
    mdv.Standard.append(ODM.Standard(
        OID="STD.SDTM.1",
        Name="SDTMIG",
        Type="IG",
        Version="3.4",
        PublishingSet="CDISC",
        Status="Final"
    ))
    
    mdv.Standard.append(ODM.Standard(
        OID="STD.CT.1",
        Name="SDTM",
        Type="CT",
        Version="2024-09-27",
        PublishingSet="CDISC",
        Status="Final"
    ))
    
    study.MetaDataVersion.append(mdv)
    odm.Study.append(study)
    
    return odm, mdv

def create_rs_domain(mdv):
    """Create RS (Disease Response) domain definition"""
    
    # Define RS domain
    rs_domain = ODM.ItemGroupDef(
        OID="IG.RS",
        Name="RS",
        Repeating="Yes",
        IsReferenceData="No",
        SASDatasetName="RS",
        Domain="RS",
        Purpose="Tabulation",
        Structure="One record per tumor response assessment per subject",
        DomainKeys="STUDYID, USUBJID, RSTESTCD, RSSEQ",
        Label="Disease Response"
    )
    
    # Add class and comment
    rs_domain.Class = ODM.Class(
        Name="Findings"
    )
    
    # Define RS variables with CDISC standards
    rs_variables = [
        {
            "name": "STUDYID",
            "label": "Study Identifier",
            "datatype": "text",
            "length": 12,
            "core": "Req",
            "origin": "Assigned",
            "comment": "Unique identifier for the study"
        },
        {
            "name": "DOMAIN",
            "label": "Domain Abbreviation",
            "datatype": "text",
            "length": 2,
            "core": "Req",
            "origin": "Assigned",
            "comment": "Two-character abbreviation for the domain (RS)"
        },
        {
            "name": "USUBJID",
            "label": "Unique Subject Identifier",
            "datatype": "text",
            "length": 40,
            "core": "Req",
            "origin": "Assigned",
            "comment": "Identifier used to uniquely identify a subject across all studies"
        },
        {
            "name": "RSSEQ",
            "label": "Sequence Number",
            "datatype": "integer",
            "length": 8,
            "core": "Req",
            "origin": "Assigned",
            "comment": "Sequence number for ordering records within a subject"
        },
        {
            "name": "RSTESTCD",
            "label": "Response Assessment Short Name",
            "datatype": "text",
            "length": 8,
            "core": "Req",
            "origin": "Assigned",
            "comment": "Short name for tumor response assessment (e.g., TUMIDENT, OVR)"
        },
        {
            "name": "RSTEST",
            "label": "Response Assessment Name",
            "datatype": "text",
            "length": 40,
            "core": "Req",
            "origin": "Assigned",
            "comment": "Full name of tumor response assessment"
        },
        {
            "name": "RSCAT",
            "label": "Category for Response",
            "datatype": "text",
            "length": 40,
            "core": "Perm",
            "origin": "Assigned",
            "comment": "Category grouping for response assessments (e.g., RECIST 1.1)"
        },
        {
            "name": "RSORRES",
            "label": "Result in Original Units",
            "datatype": "text",
            "length": 200,
            "core": "Exp",
            "origin": "CRF",
            "comment": "Result as collected on CRF or EDC"
        },
        {
            "name": "RSSTRESC",
            "label": "Character Result/Finding in Std Format",
            "datatype": "text",
            "length": 200,
            "core": "Exp",
            "origin": "Derived",
            "comment": "Standardized result (CR, PR, SD, PD for RECIST)"
        },
        {
            "name": "RSSTRESN",
            "label": "Numeric Result/Finding in Standard Units",
            "datatype": "float",
            "length": 8,
            "core": "Exp",
            "origin": "Derived",
            "comment": "Numeric representation of result if applicable"
        },
        {
            "name": "VISITNUM",
            "label": "Visit Number",
            "datatype": "integer",
            "length": 8,
            "core": "Exp",
            "origin": "Assigned",
            "comment": "Numeric identifier for visit"
        },
        {
            "name": "VISIT",
            "label": "Visit Name",
            "datatype": "text",
            "length": 40,
            "core": "Perm",
            "origin": "Assigned",
            "comment": "Protocol-defined description of visit"
        },
        {
            "name": "RSDTC",
            "label": "Date/Time of Response Assessment",
            "datatype": "datetime",
            "length": 20,
            "core": "Exp",
            "origin": "CRF",
            "comment": "Date/time of assessment in ISO8601 format"
        },
        {
            "name": "RSDY",
            "label": "Study Day of Response Assessment",
            "datatype": "integer",
            "length": 8,
            "core": "Perm",
            "origin": "Derived",
            "comment": "Study day relative to reference start date"
        }
    ]
    
    # Create ItemDefs and add to domain
    for idx, var in enumerate(rs_variables, start=1):
        # Create ItemDef
        item_def = ODM.ItemDef(
            OID=f"IT.RS.{var['name']}",
            Name=var['name'],
            DataType=var['datatype'],
            Length=var['length'],
            SASFieldName=var['name'],
            SDSVarName=var['name'],
            Origin=ODM.Origin(
                Type=var['origin']
            ),
            Label=var['label'],
            Comment=var['comment']
        )
        
        # Add to metadata version
        mdv.ItemDef.append(item_def)
        
        # Create ItemRef for domain
        item_ref = ODM.ItemRef(
            ItemOID=f"IT.RS.{var['name']}",
            Mandatory="Yes" if var['core'] == "Req" else "No",
            OrderNumber=idx,
            Role=var['core']
        )
        
        rs_domain.ItemRef.append(item_ref)
    
    # Add domain to metadata version
    mdv.ItemGroupDef.append(rs_domain)
    
    return rs_domain

def generate_define_xml_for_recist():
    """Main function to generate Define-XML for RECIST study"""
    
    print("\n🔧 Generating Define-XML v2.1 with odmlib...")
    print("=" * 60)
    
    # Create study metadata
    odm, mdv = create_study_metadata()
    print(f"✓ Created ODM root: {odm.FileOID}")
    print(f"✓ Study: {odm.Study[0].StudyName}")
    
    # Create RS domain
    rs_domain = create_rs_domain(mdv)
    print(f"\n✓ Created RS domain with {len(rs_domain.ItemRef)} variables")
    
    # Create output directory
    output_dir = Path("outputs/specs")
    output_dir.mkdir(parents=True, exist_ok=True)
    
    # Generate XML output
    xml_path = output_dir / "define_rs.xml"
    try:
        with open(xml_path, 'w', encoding='utf-8') as f:
            f.write(odm.to_xml())
        print(f"\n✅ Define-XML generated: {xml_path}")
        print(f"   Size: {xml_path.stat().st_size:,} bytes")
    except Exception as e:
        print(f"\n❌ XML generation failed: {e}")
        return False
    
    # Generate JSON output
    json_path = output_dir / "define_rs.json"
    try:
        with open(json_path, 'w', encoding='utf-8') as f:
            json.dump(odm.to_dict(), f, indent=2)
        print(f"✅ Define-JSON generated: {json_path}")
        print(f"   Size: {json_path.stat().st_size:,} bytes")
    except Exception as e:
        print(f"⚠️  JSON generation warning: {e}")
    
    # Summary
    print("\n" + "=" * 60)
    print("📊 Define-XML Summary:")
    print(f"   - Standard: SDTM IG v3.4")
    print(f"   - Controlled Terminology: 2024-09-27")
    print(f"   - Domains: RS (Disease Response)")
    print(f"   - Variables: {len(rs_domain.ItemRef)}")
    print(f"   - Format: Define-XML v2.1")
    print(f"   - Generated by: odmlib + pharmaverse")
    print("\n✅ Define-XML generation complete!\n")
    
    return True

if __name__ == "__main__":
    try:
        success = generate_define_xml_for_recist()
        sys.exit(0 if success else 1)
    except Exception as e:
        print(f"\n❌ Fatal error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)
